src/tcache.c

```diff
@@ -406,6 +429,49 @@ tsd_tcache_enabled_data_init(tsd_t *tsd) {
 	return false;
 }
 
+static bool
+tcache_bin_lowbits_overflowable(void *ptr, uint32_t distance) {
+	uint32_t lowbits = (uint32_t)((uintptr_t)ptr + distance);
+	return lowbits < (uint32_t)(uintptr_t)ptr;
+}
+
+static bool
+tcache_bin_init(cache_bin_t *bin, szind_t ind, uintptr_t *stack_cur) {
+	cassert(sizeof(bin->cur_ptr) == sizeof(void *));
+	/*
+	 * The full_position points to the lowest available space.  Allocations
+	 * will access the slots toward higher addresses (for the benefit of
+	 * adjacent prefetch).
+	 */
+	void *full_position = (void *)*stack_cur;
+	/*
+	 * Check with max_offset intentionally.  It is more conservative than
+	 * necessary (bin_stack_size + 1 ptr), where the extra 1 is for
+	 * low_water (and is easy to miss).
+	 */
+	if (tcache_bin_lowbits_overflowable(full_position,
```

Comment:
Why do we need to check for overflow?

It seems like we can do an equality check on the low-water mark (and change the meaning of low-water slightly) instead. The current strategy wastes tcache space, right?

Reply:
The pointer compression (for low_water and full_position) needs the guarantee that the space do not cross the 32-bit / 4G boundary.

When we allocate the total space for the stacks (https://github.com/jemalloc/jemalloc/pull/1592/files#diff-777e740de5d78972a2742086c8dc7f67R530), we check if the low bits are overflowable and if so add some padding. And then when initializing the individual stacks, add the padding for the specific stack -- so yep that tcache stack space is wasted in that case. It rarely happens though.

re: checking with low-water vs max_offset -- I intentionally made this conservative with max_offset. It's true that stack_size + sizeof(ptr) (i.e. the limit of low_water) is sufficient; however in case we change the limit in the future, it's easy to miss the change here (plus we don't really save much from it).

Reply:
> The pointer compression (for low_water and full_position) needs the guarantee that the space do not cross the 32-bit / 4G boundary.

What I mean is, if we make it an equality check, we don't need that guarantee (right?). I.e. we can make the algorithm resilient to overflow, so we don't need to allocate extra padding.

Reply:
Actually, come to think of it, any reason why we can't align the cache bin allocation so that overflow is impossible to begin with?

Reply:
> What I mean is, if we make it an equality check, we don't need that guarantee (right?). I.e. we can make the algorithm resilient to overflow, so we don't need to allocate extra padding.

You mean we always check with `== low_water`? That should be possible; but does feel pretty risky though.

> Actually, come to think of it, any reason why we can't align the cache bin allocation so that overflow is impossible to begin with?

Good point. What do you think about, when allocating the total stack space, align with the next power of 2 ceiling of total_stack_size?

Reply:
Just updated: added alignment to avoid low bits overflow.